<a href="https://colab.research.google.com/github/Mephod1y/DataScience/blob/main/Hw7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
pip install surprise

In [4]:
import pandas as pd
import numpy as np
from surprise import SVD, SVDpp, NMF
from surprise import Dataset
from surprise.model_selection import cross_validate

In [6]:
data = Dataset.load_builtin('ml-100k')

Dataset ml-100k could not be found. Do you want to download it? [Y/n] y
Trying to download dataset from https://files.grouplens.org/datasets/movielens/ml-100k.zip...
Done! Dataset ml-100k has been saved to /root/.surprise_data/ml-100k


In [7]:
algorithms = [SVD(), SVDpp(), NMF()]
results = []
for algorithm in algorithms:
  result = cross_validate(algorithm, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)
  results.append((algorithm, result))
  print(result)

Evaluating RMSE, MAE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9360  0.9369  0.9385  0.9362  0.9326  0.9360  0.0019  
MAE (testset)     0.7391  0.7356  0.7408  0.7361  0.7352  0.7374  0.0022  
Fit time          0.96    1.00    1.06    0.98    1.01    1.00    0.04    
Test time         0.17    0.17    0.11    0.23    0.11    0.16    0.04    
{'test_rmse': array([0.93595075, 0.9369101 , 0.93845661, 0.93623387, 0.93261003]), 'test_mae': array([0.73907501, 0.73559678, 0.74080443, 0.73610896, 0.73523576]), 'fit_time': (0.9551172256469727, 1.0011544227600098, 1.06355881690979, 0.9802589416503906, 1.0100209712982178), 'test_time': (0.1679058074951172, 0.17468953132629395, 0.10620999336242676, 0.2257096767425537, 0.10815715789794922)}
Evaluating RMSE, MAE of algorithm SVDpp on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9191  0.9159  0.9207  0.91

In [10]:
for algorithm, result in results:
    print(type(algorithm).__name__)
    print('test_rmse:', np.mean(result['test_rmse']).round(5))
    print('test_mae', np.mean(result['test_mae']).round(5))
    print('fit_time:', np.mean(result['fit_time']).round(5))
    print('test_time:', np.mean(result['test_time']).round(5))


SVD
test_rmse: 0.93603
test_mae 0.73736
fit_time: 1.00202
test_time: 0.15653
SVDpp
test_rmse: 0.91918
test_mae 0.72147
fit_time: 22.37062
test_time: 3.88771
NMF
test_rmse: 0.96401
test_mae 0.75773
fit_time: 1.67823
test_time: 0.15526


SVD - має середню точність 0.737, але дуже швидке обчислення. Доцільно використовувати коли датасети дуже великі і точність не має принципового значення.

SVD++ - має найкращу точність 0.721, але дуже повільне обчислення в порівнянні з SVD. Варто використовувати коли найкраща точність має значення.

NMF - має найгіршу точність з перевірених алгоритмів, але швидкість навчання трішки майже як в SVD.

Краще використовувати SVD алгоритм мінімальні затрати ресурсів з приємлемим результатом, але коли результат повинен бути максимально точний краще використовувати SVD++.